File: P2-ListADT-c/Lex.c
//-----------------------------------------------------------------------------
// Lex.c
// Kevin Duong
// keduong
// 1-27-17
// Sorts given in file using List ADT into out file in lex order.
// Used FileIO.c as framework
//-----------------------------------------------------------------------------
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 160

int main(int argc, char * argv[]){

   int count=0;
   int i, j;
   FILE *in, *out;
   char line[MAX_LEN];
 

   // check command line for correct number of arguments
   if( argc != 3 ){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if( out==NULL ){
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }

   /* read each line of input file, then count and print tokens */
   while( fgets(line, MAX_LEN, in) != NULL)  {
      count++;
   }
   
   /* return to start */
   rewind(in);
   
   /* create string array and loop through */
   char strarry[count-1][MAX_LEN];
   int lines = -1;
   while(fgets(line,MAX_LEN,in) != NULL){
      strcpy(strarry[++lines], line);
   }
   
   /* make list and sort */
   List L = newList();
   char* element;
   append(L,0);
   for(i=1; i<count; ++i){
      element = strarry[i];
      j = i-1;
      moveFront(L);
      while(j>=0 && strcmp(element, strarry[get(L)])<=0){
         --j;
         moveNext(L);
      }
   if(index(L)>=0) 
      insertBefore(L,i);
   else
      append(L,i);  
   }   
   
   /* Return to front and print*/
   moveBack(L);
   while(index(L)>=0){
      fprintf(out, "%s", strarry[get(L)]);
      movePrev(L);
   }
   
   /* close files and free list*/
   fclose(in);
   fclose(out);
   freeList(&L);
   
   return(0);
}

File: P1-ListADT-c/Lex.java
// Lex.java
// Kevin Duong 
// keduong
// 1-17-17
// PA1 - Lex.java
// Sorts given in file using List ADT into out file in lex order.
// Used FileIO.java as framework

import java.io.*;
import java.util.Scanner;

class Lex{
   public static void main(String[] args) throws IOException{
      
      // variables --------------------------------------------
      Scanner in = null;
      PrintWriter out = null;
      String line = null;
      String[] token = null;
      int i, j, n=0;
      int lineNumber = -1;

      // input check -------------------------------------------
      if(args.length != 2){
         System.err.println("Usage: Lex infile outfile");
         System.exit(1);
      }
      
      // count number of lines n in the file and create String array
      in = new Scanner(new File(args[0]));
      out = new PrintWriter(new FileWriter(args[1]));      
      while( in.hasNextLine() ){
         in.nextLine();
         ++n;

      }
      token = new String[n];
      
      // remake scanner for new string array and list --------------------    
      in.close();
      in = null;    
      in = new Scanner(new File(args[0]));
      
      // make string array and put lines in--------------------------------
      while(in.hasNextLine()){
         ++lineNumber;
         token[lineNumber] = in.nextLine();
      }
      // new list for integers (indices)-----------------------------------      
      List L = new List();
      L.append(0); // firstline
      
      // insertion sort algorithm *for indicies, not elements*
      for (i=1;i<token.length;++i){ 
         String s = token[i];
         j = i-1; // loop control
         L.moveFront(); // find the appropriate index
         while(j>=0 && s.compareTo(token[L.get()]) <=0){ //string comparison
            L.moveNext();
            --j;
         }
         if (L.index() >=0) L.insertBefore(i); // insert in the list in there
         else L.append(i); // we fell off so need to append
      }
      
      // printing ------------------------------------------------------------
      L.moveBack(); // get the cursor under the back element
      while(L.index() >= 0){
         out.println(token[L.get()]); 
         L.movePrev(); // go all the way through the list
      }
      in.close();
      out.close();
   }
}